cli/pack/app_dir.go

```diff
@@ -106,48 +135,142 @@ func copyFromCache(paths CachePaths, destPath string, ctx *context.Ctx) {
 	}
 
-	rockspecPath, err := common.FindRockspec(ctx.Project.Path)
+	fileContent, err := common.GetFileContentBytes(cacheParamsPath)
 	if err != nil {
-		return nil, fmt.Errorf("Unable to find rockspec: %s", err)
-	} else if rockspecPath == "" {
-		return nil, fmt.Errorf("Application directory should contain rockspec")
+		return nil, fmt.Errorf("Failed to read %s file: %s", cacheParamsPath, err)
+	}
+
+	var cachePathsParams CachePathsParams
+	if err := yaml.Unmarshal(fileContent, &cachePathsParams); err != nil {
+		return nil, fmt.Errorf("Failed to parse file %s with pack cache parameters: %s", cacheParamsPath, err)
+	}
+
+	return cachePathsParams, nil
+}
+
+func calculateCachePath(ctx *context.Ctx, params *CachePathParams) (string, error) {
+	var keyHash string
+	projectPathHash := common.StringSHA1Hex(ctx.Project.Path)[:cntFirstSymbolsFromHash]
+
+	switch {
+	case params.AlwaysCache == true:
+		keyHash = "always"
+	case params.KeyPath != "":
+		pathFromProjectRoot := filepath.Join(ctx.Project.Path, params.KeyPath)
+		if _, err := os.Stat(pathFromProjectRoot); err == nil {
+			if keyHash, err = common.FileSHA1Hex(pathFromProjectRoot); err != nil {
+				return "", fmt.Errorf("Failed to get hash from file content for path %s: %s", params.Path, err)
+			}
+		} else if os.IsNotExist(err) {
+			return "", fmt.Errorf("Specified key-path file %s for the path %s does not exist", pathFromProjectRoot, params.Path)
+		} else {
+			return "", fmt.Errorf("Failed to get specified file for path %s: %s", params.Path, err)
+		}
+
+		keyHash = keyHash[:cntFirstSymbolsFromHash]
+	default:
+		keyHash = common.StringSHA1Hex(params.Key)[:cntFirstSymbolsFromHash]
+	}
+
+	return filepath.Join(ctx.Cli.CacheDir, projectPathHash, params.Path, keyHash), nil
+}
+
+func validateCacheParams(params *CachePathParams) error {
+	if params.Key != "" && params.KeyPath != "" {
+		return fmt.Errorf(cacheParamsErrorMsg, params.Path)
+	}
+
+	if params.AlwaysCache == true && (params.Key != "" || params.KeyPath != "") {
+		return fmt.Errorf(cacheParamsErrorMsg, params.Path)
+	}
+
+	if params.AlwaysCache == false && params.Key == "" && params.KeyPath == "" {
+		return fmt.Errorf(cacheParamsErrorMsg, params.Path)
 	}
 
-	projectPathHash := common.StringSHA1Hex(ctx.Project.Path)[:10]
-	rockspecHash, err := common.FileSHA1Hex(rockspecPath)
+	return nil
+}
+
+func getProjectCachePaths(ctx *context.Ctx) (CachePaths, error) {
+	if ctx.Pack.NoCache {
+		return nil, nil
+	}
+
+	cachePathsParams, err := parseCacheParamsFile(filepath.Join(ctx.Project.Path, cacheParamsFileName))
 	if err != nil {
 		return nil, err
 	}
 
-	rockspecHash = rockspecHash[:10]
+	cachePaths := CachePaths{}
+	for _, params := range cachePathsParams {
+		if err := validateCacheParams(&params); err != nil {
+			return nil, err
+		}
+
+		cachePath, err := calculateCachePath(ctx, &params)
+		if err != nil {
+			return nil, err
+		}
+
+		cachePaths[params.Path] = cachePath
```

Comment:
What if some path passed twice?

Reply:
Added check and error in case of duplicate paths